[ML] Report correct count for df-analytics get-stats API

dimitris-athanasiou opened the pull request:
The count should match the number of all df-analytics that
matched the id in the request. However, we used to set the count
to the number of df-analytics returned which was bound to the
`size` parameter.

elasticmachine commented:
Pinging @elastic/ml-core

dimitris-athanasiou commented:
run elasticsearch-ci/packaging-sample

dimitris-athanasiou commented on x-pack/plugin/src/test/resources/rest-api-spec/test/ml/data_frame_analytics_crud.yml:
```diff
@@ -802,14 +802,14 @@ setup:
         size: 2
-  - match: { count: 2 }
+  - match: { count: 3 }
   - match: { data_frame_analytics.0.id: "bar" }
```
I want to specifically assert the `count` field rather than the objects returned.

davidkyle replied:
I mean in addition to. So in this case because `size: 2`
```
- match: { count: 3 }
- length: { data_frame_analytics : 2 }

dimitris-athanasiou replied:
Ah, got it. Sounds good.

dimitris-athanasiou commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/GetDataFrameAnalyticsStatsAction.java:
```diff
@@ -53,12 +53,14 @@ private GetDataFrameAnalyticsStatsAction() {
-        private String id;
-        private boolean allowNoMatch = true;
-        private PageParams pageParams = PageParams.defaultParams();
+        private volatile String id;
```
We set things on that request object on the callbacks of async calls. As far as I understand those could happen on a different thread?

davidkyle commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java:
```diff
@@ -137,9 +137,10 @@ protected void doExecute(Task task, GetDataFrameAnalyticsStatsAction.Request req
                 request.setExpandedIds(expandedIds);
                 ActionListener<GetDataFrameAnalyticsStatsAction.Response> runningTasksStatsListener = ActionListener.wrap(
-                    runningTasksStatsResponse -> gatherStatsForStoppedTasks(request.getExpandedIds(), runningTasksStatsResponse, listener),
+                    runningTasksStatsResponse -> gatherStatsForStoppedTasks(request, runningTasksStatsResponse, listener),
```
Rather than adding a `count` field to the request -which feels odd because it is part of the response not a request param-  you could use `response.getResources().count()` in the `gatherStatsForStoppedTasks` listener

```suggestion
                    runningTasksStatsResponse -> gatherStatsForStoppedTasks(request, runningTasksStatsResponse, ActionListener.wrap(finalResponse -> { listener.onResponse(new Response(finalResponse.getStats(), response.getResources().count() ),
                    ...
```

dimitris-athanasiou replied:
Ah, that is a very good point. I'll do that.

dimitris-athanasiou committed:
[ML] Report correct count for df-analytics get-stats API

The count should match the number of all df-analytics that
matched the id in the request. However, we set the count
to the number of df-analytics returned which was bound to the
`size` parameter. This commit fixes this by setting the count
to the count of the `get` response.

dimitris-athanasiou committed:
Also assert length of returned objects

davidkyle reviewed (approved):
LGTM
